In [1]:
import pandas as pd
from sklearn.tree import (
    DecisionTreeClassifier,
    DecisionTreeRegressor,
    export_text
)
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, mean_absolute_error
from sklearn.ensemble import (
    RandomForestClassifier,
    RandomForestRegressor)
from xgboost import XGBClassifier, XGBRegressor
import matplotlib.pyplot as plt
from sklearn.model_selection import cross_val_score

In [3]:
df = pd.read_csv("heart_disease.csv")

# Task 1: Load and Prepare the Data

In [5]:
X = df.drop(columns="target")
y = df['target']

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)


# Task 2: Train a Single Decision Tree

In [11]:
dt = DecisionTreeClassifier(random_state=42)
dt.fit(X_train, y_train)

dt_train_acc = accuracy_score(y_train, dt.predict(X_train))
dt_test_acc = accuracy_score(y_test, dt.predict(X_test))

print("Decision Tree train accuracy: ", round(dt_train_acc,3))
print("Decision Tree test accuracy:  ", round(dt_test_acc, 3))

Decision Tree train accuracy:  1.0
Decision Tree test accuracy:   0.685


-  Training accuracy is very good at predicting information it has seen before but has trouble when introduced to new data. 

-  The model is overfit

# Task 3: Visualize the Decision Tree

In [18]:
X.columns

Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach',
       'exang', 'oldpeak', 'slope', 'ca', 'thal'],
      dtype='object')

In [25]:
feature_names = ['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal']

print(export_text(
    dt,
    feature_names=feature_names

))

print(f" leaves = {dt.get_n_leaves()}")
print(f" depth = {dt.get_depth()}")

|--- cp <= 2.50
|   |--- oldpeak <= 2.45
|   |   |--- thal <= 1.50
|   |   |   |--- chol <= 272.00
|   |   |   |   |--- trestbps <= 109.00
|   |   |   |   |   |--- chol <= 232.50
|   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |--- chol >  232.50
|   |   |   |   |   |   |--- class: 1
|   |   |   |   |--- trestbps >  109.00
|   |   |   |   |   |--- class: 0
|   |   |   |--- chol >  272.00
|   |   |   |   |--- oldpeak <= 1.35
|   |   |   |   |   |--- thalach <= 138.00
|   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |--- thalach >  138.00
|   |   |   |   |   |   |--- chol <= 330.00
|   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |--- chol >  330.00
|   |   |   |   |   |   |   |--- chol <= 337.50
|   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |--- chol >  337.50
|   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |--- oldpeak >  1.35
|   |   |   |   |   |--- class: 1
|   |   |--- thal >  1.50
|   |   |   |--- thalac

-  depth: 10
-  leaf node count: 36
-  first question: "cp <= 2.50

# Task 4: Train a Random Forest

In [28]:
rf = RandomForestClassifier(n_estimators=100, random_state=42)
rf.fit(X_train, y_train)

rf_train_acc = accuracy_score(y_train, rf.predict(X_train))
rf_test_acc = accuracy_score(y_test, rf.predict(X_test))

print(f"RF Training Accuracy: {rf_train_acc}")
print(f"RF Test Accuracy: {rf_test_acc}")

print("Decision Tree train accuracy: ", round(dt_train_acc,3))
print("Decision Tree test accuracy:  ", round(dt_test_acc, 3))

RF Training Accuracy: 1.0
RF Test Accuracy: 0.7962962962962963
Decision Tree train accuracy:  1.0
Decision Tree test accuracy:   0.685


-  Test accuracy improved on the Random Forest Model by ~10%
-  The gap did get smaller, this suggests that overfitting wasn't as prominent as it was in the Decision Tree model.

# Task 5: Train an XGBoost Model

# Task 6: Compare All Three Models

| Model           | Train Accuracy | Test Accuracy |
|-----------------|----------------|---------------|
| Decision Tree   |                |               |
| Random Forest   |                |               |
| XGBoost         |                |               |